File: backend/app/domain/services/audit.py
import uuid
from datetime import datetime, timezone
from typing import Any, Optional

from sqlalchemy.orm import Session

from app.domain.entities.system import AuditLog


def log_audit_action(
    session: Session,
    action: str,
    entity_type: str,
    actor_id: Optional[uuid.UUID] = None,
    entity_id: Optional[str] = None,
    details: Optional[dict[str, Any]] = None
) -> AuditLog:
    """Creates and persists an append-only audit log record."""
    audit_entry = AuditLog(
        id=uuid.uuid4(),
        actor_id=actor_id,
        action=action,
        entity_type=entity_type,
        entity_id=entity_id,
        details=details or {},
        created_at=datetime.now(timezone.utc)
    )
    session.add(audit_entry)
    session.flush()  # flush to assign IDs without committing the whole transaction yet
    return audit_entry

File: backend/src/lib/audit.ts
import { v4 as uuidv4 } from "uuid";
import type { Prisma } from "../generated/prisma";

type Tx = Prisma.TransactionClient;

/** Port of app/core/services/audit.py log_audit_action. */
export async function logAuditAction(
  tx: Tx,
  action: string,
  entityType: string,
  actorId?: string | null,
  entityId?: string | null,
  details?: Record<string, unknown>,
): Promise<void> {
  await tx.auditLog.create({
    data: {
      id: uuidv4(),
      actorId: actorId ?? null,
      action,
      entityType,
      entityId: entityId ?? null,
      details: (details ?? {}) as Prisma.InputJsonValue,
      createdAt: new Date(),
    },
  });
}
